### PR TITLE
wordy: Generate tests

### PR DESCRIPTION
**Pull request by tommyschaefer**
Hi there!

I've made a first pass at converting the `wordy` exercise to use generated tests (#396). 

I based the updates on #474, but I still have a few questions:

1. The original tests use a mixture of the following forms:

```
question = '...'
assert_equal ...
```
and
```
assert_equal 1, WordProblem.new('...').answer
```

I'm assuming this has to do with not going past 80 characters per line. In my first pass I decided to just use the first form for all of the tests. Is this okay, or should I put some work into switching between the two forms based on the length of the assertion?

2. Does anything special have to happen since this wasn't originally book-kept?

3. In the `test_add_then_multiply` / `test_addition_and_multiplication` test, the xruby implementation includes a custom error message. Should this error message be removed (as it is currently), or should we add the custom error message to the shared metadata and generate based on that?

**Thank you so much for your time :smile: !**

### CODE REVIEW

File: exercises/wordy/wordy_test.rb [hunk 1]
```diff
@@ -3,95 +3,127 @@
...
     skip
     question = 'What is 2 multiplied by -2 multiplied by 3?'
     assert_equal(-12, WordProblem.new(question).answer)
   end
 
-  def test_add_then_multiply
+  def test_addition_and_multiplication
     skip
     question = 'What is -3 plus 7 multiplied by -2?'
-    message = 'You should ignore order of precedence. -3 + 7 * -2 = -8, not -17'
```
Comment (tommyschaefer) [inline exercises/wordy/wordy_test.rb]: This is the custom message that isn't currently generated.

File: exercises/wordy/example.tt [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env ruby
```
Comment (tommyschaefer) [inline exercises/wordy/example.tt]: I know that there was some discussion about removing this in #295. Should it be included in newly generated tests? I wasn't too sure whether a decision had been made 😄

Comment (Insti) [inline exercises/wordy/example.tt]: I don't think a conclusion was ever reached, you could try leaving it out and see if anyone notices.

File: exercises/wordy/example.tt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+gem 'minitest', '>= 5.0.0'
+require 'minitest/autorun'
+require_relative 'wordy'
+
+# Test data version:
+# <%= sha1 %>
```
Comment (Insti) [inline exercises/wordy/example.tt]: This should be one line.

File: exercises/wordy/wordy_test.rb [hunk 4]
```diff
@@ -3,95 +3,127 @@
...
+    question = 'What is 53 plus 2?'
+    assert_equal 55, WordProblem.new(question).answer
   end
 
-  def test_add_negative_numbers
+  def test_addition_with_negative_numbers
     skip
-    assert_equal(-11, WordProblem.new('What is -1 plus -10?').answer)
+    question = 'What is -1 plus -10?'
+    assert_equal(-11, WordProblem.new(question).answer)
```
Comment (Insti) [inline exercises/wordy/wordy_test.rb]: Why does this one have brackets around the arguments?

**Comment by Insti** [comment]
> The original tests use a mixture of the following forms:
> (snip)
>  I decided to just use the first form for all of the tests.

This is a good decision.

(Need to step out now, will continue review a bit later.)

File: exercises/wordy/wordy_test.rb [hunk 5]
```diff
@@ -3,95 +3,127 @@
...
     assert_raises ArgumentError do
-      WordProblem.new('What is 53 cubed?').answer
+      WordProblem.new(question).answer
     end
   end
 
-  def test_irrelevant
+  def test_non_math_question
     skip
+    question = 'Who is the President of the United States?'
```
Comment (Insti) [inline exercises/wordy/wordy_test.rb]: This question doesn't start with 'What is' but that's probably an issue for the common metadata.

**Comment by Insti** [comment]
@tommyschaefer  you might be interested in looking at my proposal for how test generators should work in future: https://github.com/Insti/xruby/pull/1

File: exercises/wordy/wordy_test.rb (hunk 4 above)
Comment (tommyschaefer) [inline exercises/wordy/wordy_test.rb]: Hi there!

I originally thought that it was something enforced by rubocop, but I'm not sure now. Whenever I save the file without the parentheses, I get a warning in Vim `ambiguous first argument; put parentheses or a space even after `-' operator` (only for cases with a negative number).

I'm assuming that's maybe why the original cases also had parens around negative cases?

**Comment by tommyschaefer** [comment]
@Insti I took a look and it looks great! I based the interface of the `WordyCase` class on the discussions on #463. Is there anything I can do to `WordyCases` to make it easier to convert to the future generator? The thing that jumped out most to me was the inclusion of a `workload` method, but is there anything else that would make it easier to update?

Thank you so much for your help!

**Comment by Insti** [comment]
Don't worry too much about making this future compatible, the new format is not finalized and there will be a project to review all the generators when it is.

File: exercises/wordy/wordy_test.rb (hunk 4 above)
Comment (Insti) [inline exercises/wordy/wordy_test.rb]: I think that's a Rubocop warning, it's not particularly important, but it's better to make all the cases consistent so in this case I'd wrap them all in parentheses. This has the pleasant side effect of making your `value_assertion` method much simpler.

File: exercises/wordy/wordy_test.rb (hunk 1 above)
Comment (Insti) [inline exercises/wordy/wordy_test.rb]: Can you work out how to make the generator add this message yourself, or would you like some hints?

File: lib/wordy_cases.rb [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+      "assert_raises ArgumentError do",
+      indent(2, "WordProblem.new(question).answer"),
+      "end",
+    ].join("\n")
+  end
+end
+
+WordyCases = proc do |data|
+  JSON.parse(data)['cases'].map.with_index do |row, i|
+    WordyCase.new(row.merge('index' => i))
```
Comment (Insti) [inline lib/wordy_cases.rb]: You can preprocess the json here to add a failure message:
```
wordy_case = WordyCase.new(row.merge('index' => i, failure_message => nil))
if wordy_case.question = 'What is -3 plus 7 multiplied by -2?'
  wordy_case.failure_message = 'You should ignore order of precedence. -3 + 7 * -2 = -8, not -17'
end
wordy_case
```
(untested, a Github comment box is not an ideal programming environment.)

File: lib/wordy_cases.rb [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  def indent(size, lines)
+    lines.split("\n").map do |line|
+      ' ' * size + line
+    end.join("\n")
+  end
+
+  def assertion
+    return error_assertion unless expected
+
+    "assert_equal(#{expected}, WordProblem.new(question).answer)"
```
Comment (Insti) [inline lib/wordy_cases.rb]: Then you can add your custom failure message here.
```
message = failure_message || "some generic but helpful error message"
"assert_equal(#{expected}, WordProblem.new(question).answer, #{message})"
```

File: exercises/wordy/wordy_test.rb (hunk 1 above)
Comment (Insti) [inline exercises/wordy/wordy_test.rb]: See my latest comments.

File: lib/wordy_cases.rb (hunk 6 above)
Comment (Insti) [inline lib/wordy_cases.rb]: One thing to bear in mind here is that the failure message should show the actual value returned.

**Comment by tommyschaefer** [comment]
@insti Fantastic! Thank you so much for the help! I'll make these updates and push them this evening! 😀

**Comment by tommyschaefer** [comment]
@Insti Does this look ready to review to you? If so, would you like me to go ahead and squash the commits? Thank you so much for your time and help!

**Comment by kotp** [comment]
This looks good.  You could squash.

**Comment by tommyschaefer** [comment]
Hmm... I'm not sure why this is failing all of the sudden. I don't think its based on these commits.

Relevant output of the log:
```
$ export BUNDLE_GEMFILE=$PWD/Gemfile
$ ruby --version
ruby 2.1.2p95 (2014-05-08 revision 45877) [x86_64-linux]
$ rvm --version
rvm 1.26.10 (latest-minor) by Wayne E. Seguin <wayneeseguin@gmail.com>, Michal Papis <mpapis@gmail.com> [https://rvm.io/]
$ bundle --version
Bundler version 1.6.5
$ gem --version
2.4.5
$ bundle install --quiet
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    /home/travis/.rvm/rubies/ruby-2.1.2/bin/ruby mkrf_conf.rb 

/home/travis/.rvm/rubies/ruby-2.1.2/bin/ruby -rubygems /home/travis/.rvm/gems/ruby-2.1.2/gems/rake-12.0.0/exe/rake RUBYARCHDIR=/home/travis/.rvm/gems/ruby-2.1.2/extensions/x86_64-linux/2.1.0/rainbow-2.2.0 RUBYLIBDIR=/home/travis/.rvm/gems/ruby-2.1.2/extensions/x86_64-linux/2.1.0/rainbow-2.2.0
(in /home/travis/.rvm/gems/ruby-2.1.2/gems/rainbow-2.2.0)
fatal: Not a git repository (or any of the parent directories): .git
rake aborted!
LoadError: cannot load such file -- rspec/core/rake_task
/home/travis/.rvm/gems/ruby-2.1.2/gems/rainbow-2.2.0/Rakefile:2:in `<top (required)>'
/home/travis/.rvm/gems/ruby-2.1.2/gems/rake-12.0.0/exe/rake:27:in `<main>'
(See full trace by running task with --trace)

rake failed, exit code 1
```

**Comment by kotp** [comment]
The configuration shows that rvm is version 2.1.2.  I don't think this is correct..
```
rvm:
    version:      "rvm 1.28.0 (master) by Wayne E. Seguin <wayneeseguin@gmail.com>, Michal Papis <mpapis@gmail.com> [https://rvm.io/]"
```

Shows the version of RVM that I am using that was updated only a few days ago.

**Comment by kotp** [comment]
Can you rebase on master and submit to test the travis-ci?

**Comment by tommyschaefer** [comment]
@kotp Looks like it worked! 😄

**Comment by kotp** [comment]
Ready for a squash again.

**Comment by Insti** [comment]
Thanks for all your work on this @tommyschaefer :heart:

**Pull request merged**